### PR TITLE
Feat/backend structure

### LINKED ISSUE #12: [Feat]: Initialize backend structure
### My suggestion

Create abstract structure of backend with place to populate logic later.

- [x] Landsat API Integration, DX friendly interface for api communication
- [x] Coordinate and WRS-2 System, Toolbox for Coordination lookup and Scene Detection functions
- [x] Notification Service, Abstract interface for sending notifications
- [ ] Data Filtering and Processing module, interface for processing api provided data
- [x] Data export, interface for generating reports 
- [x] Frontend communication endpoints, for now example post and get should do

How it should look is up to you.

use `raise NotImplementedError` over `pass`  

### Possible alternatives

_No response_

### PR DESCRIPTION
**Pull request by wikipop**
Resolves #12

### CODE REVIEW

File: backend/src/types/__init__.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+from typing import List, Tuple
+
+bbox_type = Tuple[int, int, int, int]
```
Comment (Norbiros) [inline backend/src/types/__init__.py]:
```suggestion
BBoxType = Tuple[int, int, int, int]
```
https://docs.python.org/3/library/typing.html#type-aliases

Comment (Norbiros) [inline backend/src/types/__init__.py]: What are those ints? `x start, x end, y start, y end`? Add information about it

File: backend/src/classes/LandSatData.py [hunk 2]
```diff
@@ -0,0 +1,2 @@
+class LandSatData:
```
Comment (Norbiros) [inline backend/src/classes/LandSatData.py]: On NASA website they use `Landsat 9`, `Landsat Science`, not `LandSat`

File: backend/src/tools/__init__.py
Comment (Norbiros) [inline backend/src/tools/__init__.py:1]: I don't understand this category. Maybe it's better to call it `utils`?

File: backend/src/services/__init__.py [hunk 3]
```diff
@@ -0,0 +1,4 @@
+from NotificationService import Notification, NotificationService
+
+Notification = Notification
```
Comment (Norbiros) [inline backend/src/services/__init__.py]: Ymm, I'm not sure if this `Notification = Notification` even makes a difference. Reexports don't work like that

File: backend/src/tools/WRS2Tools.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+from datetime import datetime
+
+from backend.src.types import bbox_type
+
+class WRS2Tools:
```
Comment (Norbiros) [inline backend/src/tools/WRS2Tools.py]: Explain what WRS2 even is

File: backend/src/tools/WRS2Tools.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    @staticmethod
+    def get_wrs2_path(row, path):
+        """
+        Get the WRS-2 path number for a given row and path.
+        """
+        raise NotImplementedError
+
+    @staticmethod
+    def get_wrs2_tile(lat, lon):
```
Comment (Norbiros) [inline backend/src/tools/WRS2Tools.py]: Add types

File: backend/src/tools/WRS2Tools.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
+from datetime import datetime
+
+from backend.src.types import bbox_type
+
+class WRS2Tools:
+    @staticmethod
+    def get_wrs2_path(row, path):
```
Comment (Norbiros) [inline backend/src/tools/WRS2Tools.py]: Same as below

File: backend/src/services/ReportGeneratorService.py [hunk 7]
```diff
@@ -0,0 +1,9 @@
+from backend.src.classes import LandSatData
+
+
+class ReportGeneratorService:
+    def __init__(self, Data: LandSatData):
```
Comment (Norbiros) [inline backend/src/services/ReportGeneratorService.py]:
```suggestion
    def __init__(self, data: LandSatData):
```

File: backend/src/services/NotificationService.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+
+class Notification(Enum):
+    EMAIL = 1
+    PUSH = 2
+
+
+class NotificationService:
+    @staticmethod
+    def sendNotification(user, message, notificationType: Notification):
```
Comment (Norbiros) [inline backend/src/services/NotificationService.py]:
```suggestion
    def sendNotification(user, message, notification_type: Notification):
```

**Review by Norbiros — CHANGES_REQUESTED** [review]

File: backend/classes/LandsatAPI.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+    def __init__(self):
+        raise NotImplementedError
+
+    def search(self, bbox: BBoxType,
+               start_date: datetime,
+               end_date: datetime,
+               limit: int,
+               cloud_cover,
+               sort_by) -> [LandsatData]:
```
Comment (Norbiros) [inline backend/classes/LandsatAPI.py]: Missing types!

File: backend/routers/__init__.py [hunk 10]
```diff
@@ -0,0 +1,5 @@
+from backend.routers.coordinates_router import coordinates_router
+from backend.routers.index_router import index_router
+
+index_router = index_router
+coordinates_router = coordinates_router
```
Comment (Norbiros) [inline backend/routers/__init__.py:5]: Why?

File: backend/routers/index_router.py [hunk 11]
```diff
@@ -0,0 +1,16 @@
+import structlog
+from fastapi import APIRouter, Response
+
+logger = structlog.get_logger()
```
Comment (Norbiros) [inline backend/routers/index_router.py:4]: You really have to set logger in every router? Why not in `__init__.py`, or just make logger glboal variable

File: backend/middleware/__init__.py [hunk 12]
```diff
@@ -0,0 +1,3 @@
+from backend.middleware.logger_middleware import logger_middleware
+
+logger_middleware = logger_middleware
```
Comment (Norbiros) [inline backend/middleware/__init__.py:3]: Why?

File: backend/services/NotificationService.py [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+        match notification_type:
+            case Notification.EMAIL:
+                NotificationService.sendEmailNotification(user, message)
+                return
+
+            case Notification.PUSH:
+                NotificationService.sendPushNotification(user, message)
+                return
+
+    @staticmethod
```
Comment (Norbiros) [inline backend/services/NotificationService.py:21]: Why here you use `@staticmethod` but in ReportGeneratorService normal ones

File: backend/domain_types/__init__.py [hunk 14]
```diff
@@ -0,0 +1,5 @@
+from typing import Tuple
+
+# Bounding box type
+# Tuple of 4 integers: (x1, y1, x2, y2)
```
Comment (Norbiros) [inline backend/domain_types/__init__.py:4]: Um use doc coments or
```py
from dataclasses import dataclass

@dataclass
class BBoxType:
    x_min: int
    y_min: int
    x_max: int
    y_max: int
 ```

**Review by Norbiros — CHANGES_REQUESTED** [review]

File: backend/routers/index_router.py (hunk 11 above)
Comment (wikipop) [reply backend/routers/index_router.py:4]: I am trying to avoid recursive imports with reexporting from `__init__.py`

**Pull request merged**